peps/pep-0758.rst

```diff
@@ -0,0 +1,432 @@
+downloads for any wheel with a matching checksum.  They could also query PyPI
+for the project's organization owner, then verify that it has a support URI
+before trusting the download. They could warn the user when externally hosted
+wheel files are encountered, and/or require the use of a command line option
+to enable additional download hosts. Any of these verification policies could
+be chosen in configuration files.
+
+Installer tools should also probably provide better error messages when
+externally hosted wheels cannot be downloaded, e.g. because a host is
+unreachable. As described above, such tools could query enough metadata from
+PyPI to provide clear and distinct error messages pointing users to the
+package's external hosting support email or issue tracker.
+
+Security
+========
+
+Several factors as described in this proposal should mitigate security
+concerns with externally hosted wheels, such as:
+
+- Wheel file checksums MUST be included in ``.rim`` files, and once uploaded
+  cannot be changed. Since the checksum stored on PyPI is immutable and
+  required, it is not possible to spoof an external wheel file, even if the
+  owning organization lost control of their hosting domain.
+- Externally hosted wheels MUST be served over HTTPS.
+- In order to serve externally hosted wheels, organizations MUST be approved
+  by the PyPI admins.
+
+When users identify malware or vulnerabilities in PyPI-hosted projects, they
+can now report this using the `malware reporting facilities
+<https://pypi.org/security/>`__ on PyPI, as also described in this
+`blog post
+<https://blog.pypi.org/posts/2024-03-06-malware-reporting-evolved/>`__.  The
+same process can be used to report security issues in externally hosted
+wheels, and the same remediation process should be used.  In addition, since
+organizations with external hosting enabled MUST provide a support contact
+URI, that URI can be used in some cases to report the security issue to the
+hosting organization.  Such organization reporting won't make sense for
+malware, but could indeed be a very useful way to report security
+vulnerabilities in externally hosted wheels.
+
+Rejected ideas
+==============
+
+Several ideas were considered and rejected.
+
+- Requiring digital signatures on externally hosted wheel files, either in
+  addition to or other than hashes. We deem this unnecessary since the
+  checksum requirement should be enough to validate that the metadata on PyPI
+  for a wheel exactly matches the downloaded wheel.  The added complexity of
+  key management outweighs any additional benefit such digital signatures
+  might convey.
+- Hash verification on ``.rim`` file uploads. PyPI *could* verify that the
+  hash in the uploaded ``.rim`` file matches the externally hosted wheel
+  before it accepts the upload, but this requires downloading the external
+  wheel and performing the checksum. This would impose an ordering restriction
```

Comment:
I definitely think that wheels must be uploaded to external hosts before a RIM may be posted. If people have unpinned dependencies and a new RIM is uploaded their CI will fail.

Reply:
(I don't think this needs to be verified by PyPi however)

Reply:
Actually, I think PyPi making an HTTP HEAD request against the wheel URL is cheap and should suffice to verify that the wheel is *probably* there.

The external host could even pass the hash of the wheel in the headers of the HEAD request to verify it matches with the RIM.

If you like this idea, We should add it to the list of requirements of external hosts.

Reply:
> I definitely think that wheels must be uploaded to external hosts before a RIM may be posted. If people have unpinned dependencies and a new RIM is uploaded their CI will fail.

Agreed, although once [PEP 694](https://peps.python.org/pep-0694/) is implemented, uploading the `.rim` file to a stage before the external `.whl` is posted should be safe.

Reply:
> The external host could even pass the hash of the wheel in the headers of the HEAD request to verify it matches with the RIM.

The complication here I think is that there's no standard way to return the hash in response to the `HTTP HEAD` request.  I'm mentioning this in the rejected ideas, with a footnote added about the checksum.

Reply:
Yeah, I think just doing the HEAD request would be a significant benefit. Making sure something is there is better than nothing, and users will get a clear idea of what is wrong if the hashes mismatch at install time.